### PR TITLE
Dev krushna

### PR DESCRIPTION
**Pull request by mardikark-gslab**
This pull request contains below changes
- Updated q_install.sh supporting different modes (replacement of older aio.sh)
- UTILS/build/Makefile changes to support 'static' & 'dynamic' target
- python_q_wrapper changes

### CODE REVIEW

File: GUIDING_PRINCIPLES/build_q.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+set -e
+bash my_print.sh "STARTING: Building Q"
+pwd
+cd ../UTILS/build
+pwd
+make
```
Comment (subramon) [inline GUIDING_PRINCIPLES/build_q.sh]: might want to add a flag to make to make sure we are building with -O4 flag

File: GUIDING_PRINCIPLES/clean_up.sh [hunk 2]
```diff
@@ -1,10 +1,9 @@
...
-
-$1 $2
+#!/bin/bash
+set -e
+DIR_PATH=$1
+if [ -z $DIR_PATH ]; then
+  bash my_print.sh "No directory passed to cleanup" 1
+  exit 1
+fi
+bash my_print.sh $DIR_PATH
```
Comment (subramon) [inline GUIDING_PRINCIPLES/clean_up.sh]: whenever you expect a directory to exist, useful to do 
test -d <dir_name>
similarly, if you expect a file to exist,
test -f <file_name>

File: GUIDING_PRINCIPLES/lua_installation.sh [hunk 3]
```diff
@@ -1,22 +1,30 @@
...
+
+make linux
+sudo make install
+
+#debug flag is set then creating a link to L
+if [ "$#" -eq 1 -a $IS_LUA_DEBUG == "LUA_DEBUG" ];then
+  sudo ln -sf /usr/local/bin/lua /usr/local/bin/L
+fi
+
+sudo cp ./etc/lua.pc /usr/lib/pkgconfig/
```
Comment (subramon) [inline GUIDING_PRINCIPLES/lua_installation.sh]: what does this line do? the one with pkgconfig

File: GUIDING_PRINCIPLES/q_install.sh [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+else
+  # installing lua and luajit normal mode
+  bash lua_installation.sh
+  bash luajit_installation.sh
+fi
+
+# TODO: what's this for?
+echo "`whoami` hard nofile 102400" | sudo tee --append /etc/security/limits.conf
+echo "`whoami` soft nofile 102400" | sudo tee --append /etc/security/limits.conf
+
```
Comment (subramon) [inline GUIDING_PRINCIPLES/q_install.sh]: remember to comment these lines out and put in Indrajeet;'s notes as to why he had it in the first place

File: GUIDING_PRINCIPLES/q_required_packages.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
+bash my_print.sh "STARTING: Installing Q-required packages"
+sudo luarocks install luaposix
```
Comment (subramon) [inline GUIDING_PRINCIPLES/q_required_packages.sh]: are we using luaposix? If not, comment it out

File: GUIDING_PRINCIPLES/q_install.sh (hunk 4 above)
Comment (PragatiPailwan) [reply GUIDING_PRINCIPLES/q_install.sh]: Added Indrajeet's notes in place of above TODO in q_install.sh.

File: GUIDING_PRINCIPLES/clean_up.sh (hunk 2 above)
Comment (PragatiPailwan) [reply GUIDING_PRINCIPLES/clean_up.sh]: Now using test -d <dir_name> cmd, where a directory is expected to exist.

File: GUIDING_PRINCIPLES/q_required_packages.sh (hunk 5 above)
Comment (PragatiPailwan) [reply GUIDING_PRINCIPLES/q_required_packages.sh]: luaposix('posix.time') was used in "Q/UTILS/lua/utils.lua" file for capturing the timers.
Now we have removed the dependency for luaposix from basic pkg required for Q and added this pkg installation in "test" mode of q_install.sh as we are using it in some of the unit test files.

File: GUIDING_PRINCIPLES/lua_installation.sh (hunk 3 above)
Comment (PragatiPailwan) [reply GUIDING_PRINCIPLES/lua_installation.sh]: Referring to the man page of pkg-config:

pkg-config retrieves information about packages from special metadata files. These files are named after the package, and has a .pc extension. On  most  systems, pkg-config  looks in  /usr/lib/pkgconfig, /usr/share/pkgconfig, /usr/local/lib/pkgconfig and /usr/local/share/pkgconfig for these files.

File: GUIDING_PRINCIPLES/build_q.sh (hunk 1 above)
Comment (PragatiPailwan) [reply GUIDING_PRINCIPLES/build_q.sh]: Added the -O4 flag to QC_FLAGS for normal/basic mode of q_install.sh

**Pull request merged**